xarray/core/alignment.py

```diff
@@ -358,7 +366,15 @@ def var_indexers(var, indexers):
             args = (var.attrs, var.encoding)
         else:
             args = ()
-        reindexed[dim] = IndexVariable((dim,), indexers[dim], *args)
+
+        idx_var = IndexVariable((dim,), indexers[dim], *args)
+
+        # GH1434
+        # ensures that dtype of numpy structured arrays is preserved
+        if len(args) and is_composite_dtype(var.dtype) \
+                and idx_var.dtype != var.dtype:
+            idx_var.data = idx_var.data.astype(var.dtype)
```

Comment:
This is a simple demo that shows the problem. After concatenation dtype of the "participant" variable will be lost but it should be preserved, especially that participant is not the dimension we are concatenating along 

```python
        p1 = np.array([('A', 180), ('B', 150), ('C', 200)],
                      dtype=[('name', '|S256'), ('height', int)])
        p2 = np.array([('D', 170), ('E', 250), ('F', 150)],
                      dtype=[('name', '|S256'), ('height', int)])

        data = np.arange(50, 80, 1, dtype=np.float)

        dims = ['measurement', 'participant']

        da1 = DataArray(
            data.reshape(10, 3),
            coords={
                'measurement': np.arange(10),
                'participant': p1,
            },
            dims=dims
        )

        da2 = DataArray(
            data.reshape(10, 3),
            coords={
                'measurement': np.arange(10),
                'participant': p2,
            },
            dims=dims
        )


        combined_2 = concat([da1, da2], dim='measurement')


        assert combined_2.participant.dtype == da1.participant.dtype

```